test/blocklib/core/filter/time_domain_filter.hpp

```diff
@@ -0,0 +1,114 @@
+#ifndef GRAPH_PROTOTYPE_TIME_DOMAIN_FILTER_HPP
+#define GRAPH_PROTOTYPE_TIME_DOMAIN_FILTER_HPP
+
+#include <history_buffer.hpp>
+#include <node.hpp>
+#include <numeric>
+
+namespace gr::blocks::filter {
+
+using namespace fair::graph;
+
+/**
+ * @brief Finite Impulse Response (FIR) filter class
+ *
+ * The transfer function of an FIR filter is given by:
+ * H(z) = b[0] + b[1]*z^-1 + b[2]*z^-2 + ... + b[N]*z^-N
+ *
+ */
+template<typename T>
+    requires std::floating_point<T>
+struct fir_filter : node<fir_filter<T>> {
+    IN<T>             in;
+    OUT<T>            out;
+    std::vector<T>    b{}; // feedforward coefficients
+    history_buffer<T> inputHistory{ 32 };
+
+    void
+    init(const tag_t::map_type & /*old_settings*/, const tag_t::map_type &new_settings) noexcept {
+        if (new_settings.contains("b") && b.size() >= inputHistory.capacity()) {
+            inputHistory = history_buffer<T>(std::bit_ceil(b.size()));
+        }
+    }
+
+    constexpr T
+    process_one(T input) noexcept {
+        inputHistory.push_back(input);
+        return std::inner_product(b.begin(), b.end(), inputHistory.rbegin(), static_cast<T>(0));
+    }
+};
+
+enum class IIRForm {
+    DF_I,  /// direct form I: preferred for fixed-point arithmetics (e.g. no overflow)
+    DF_II, /// direct form II: preferred for floating-point arithmetics (less operations)
+    DF_I_TRANSPOSED,
```

Comment:
What are the transposed forms preferred for?

Reply:
Similar to the I and II but inversed. I added them more for academic/thoroughness and to give @mattkretz alternative implementations to see which one is better from a SIMD point of view.